review-problems.ml

```diff
@@ -917,3 +917,203 @@ module Lazy = struct
+  - Lists
+]
+
+In OCaml, higher-order functions are functions that can take other functions as arguments or return functions as results. 
+This question explores the concept of higher-order functions.
+
+Implement a higher-order function called `apply_to_list` that takes a function f and a list `lst`. 
+The function `apply_to_list` should apply the function `f` to each element of the list `lst` 
+and return a new list containing the results.
+
+Here are the specific tasks:
+
+- Define a higher-order function apply_to_list with the following type signature
+  val apply_to_list : ('a -> 'b) -> 'a list -> 'b list
+  The function `apply_to_list` should take a function `f` that maps values of type `'a to 'b`, and a list of values of type `'a`. 
+  It should return a list of values of type `'b`.
+
+- Write at least two example functions `f1` and `f2`, each with different type signatures. For instance, `f1` could be a function 
+  that squares an integer, and `f2` could be a function that converts a string to uppercase.
+
+- Apply the apply_to_list function to each of the example functions f1 and f2, and a list of values. 
+  Provide examples of the input list and the expected output for each function.
+
+*)
+
+
+(*
+QUESTION#4 :-
+
+Imagine you have a list of students, each represented by their name, a transcript of courses and grades
+and you want to find students who have achieved high honors. 
+
+Using higher-order functions in OCaml, solve the following problems:
+
+  - Implement a function `find_honors_student` with the following signature:
+      
+      val find_honors_student : student list -> name option
+
+      This function should take a list of students and return the name of a student who has 
+      achieved a grade of over 90 in all their courses. If there is no such student, return `None`.
+
+  - Implement a function `average_grade` with the following signature:
+
+      val average_grade : student -> float
+
+      This function should take a single student and calculate their average grade across all courses. 
+      Use this function to find the student with the highest average grade in the list of students.
+
+  - Implement a function `top_students` with the following signature:
+
+      val top_students : student list -> name list
+
+      This function should take a list of students and return a list of names of students who have achieved a 
+      grade of 95 or higher in at least one course.
+
+*)
+
+(*
+QUESTION#5 :-
+
+Recursive Chessboard Exploration
+
+Consider an `n` x `n` chessboard represented as a grid of squares. Each square can be either empty or blocked. 
+You start at the top-left square `(0, 0)` and want to reach the bottom-right square `(n-1, n-1)`. 
+However, there are some restrictions:
+
+- You can only move right or down.
+- You cannot move through blocked squares.
+- You cannot visit the same square twice.
+
+(1). Implement a recursive function `explore_chessboard` with the following signature:
```

Comment:
The idea of this question is overall good, but we have to get rid of the effectful programming concepts like arrays and mutable variables to make it a suitable question for midterm 1.

Replace the board representation from using arrays to instead use lists `bool list list`. Then, the pattern of recursion becomes very interesting because the answer to the question "is there a path from the current square to the square (n-1, n-1)" is obtained by trying to solve two subproblems: one where we drop the current _column_ from the chess board, and one where we drop the current _row_ from the chess board.

The challenge in this approach becomes how to avoid looking over the same squares over and over. The solution in that case is probably to use dynamic programming to eliminate the overlap in the solution space.
For a midterm problem that would be too challenging, since many students in 302 are only simultaneously taking the data structures and algorithms course that introduces dynamic programming.
So rather than insist on having the student write an efficient algorithm, adjust the problem to say that efficiency doesn't matter and that this algorithm is meant to be a brute-force backtracking search.

Finally, rather than just return whether whether the path exists or not, the function should return the path. This makes the backtracking search for the path more interesting because the algorithm will return `path option` where `type path = direction list` and `type direction = Down | Right`. Therefore, the results of recursive calls have to be matched on to extend the path with an additional direction.